sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -368,6 +372,66 @@ public virtual async Task<Response<UploadBlobResult>> UploadBlobAsync(Stream str
             }
         }
 
+        private async Task<ChunkedUploadResult> UploadInChunks(string location, Stream stream, int chunkSize, bool async = true, CancellationToken cancellationToken = default)
+        {
+            // TODO: don't allocate a large buffer unless we need it.  How will we know?
+            byte[] buffer = new byte[chunkSize];
+            using SHA256 sha256 = SHA256.Create();
+
+            int chunkCount = 0;
+            long blobLength = 0;
+
+            // Read first chunk into buffer.
+            int bytesRead = async ?
+                await stream.ReadAsync(buffer, 0, chunkSize, cancellationToken).ConfigureAwait(false) :
+                stream.Read(buffer, 0, chunkSize);
+
+            ResponseWithHeaders<ContainerRegistryBlobUploadChunkHeaders> uploadChunkResult = null;
+
+            while (bytesRead > 0)
+            {
+                var contentRange = GetContentRange(chunkCount * chunkSize, bytesRead);
+                location = uploadChunkResult?.Headers.Location ?? location;
+
+                // Incrementally compute hash for digest.
+                sha256.TransformBlock(buffer, 0, bytesRead, buffer, 0);
+
+                using (Stream chunk = new MemoryStream(buffer, 0, bytesRead))
+                {
+                    uploadChunkResult = async ?
+                        await _blobRestClient.UploadChunkAsync(location, chunk, contentRange, bytesRead.ToString(CultureInfo.InvariantCulture), cancellationToken).ConfigureAwait(false) :
+                        _blobRestClient.UploadChunk(location, chunk, contentRange, bytesRead.ToString(CultureInfo.InvariantCulture), cancellationToken);
+                }
+
+                blobLength += bytesRead;
+                chunkCount++;
+
+                // Read next chunk into buffer
+                bytesRead = async ?
+                    await stream.ReadAsync(buffer, 0, chunkSize, cancellationToken).ConfigureAwait(false) :
+                    stream.Read(buffer, 0, chunkSize);
+            }
```

Comment:
Have we considered wrapping stream into [BufferedStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.bufferedstream?view=net-7.0)? I was just reading these [docs](https://learn.microsoft.com/en-us/dotnet/api/system.io.bufferedstream?view=net-7.0#remarks) and seems like we can optimize read performance using BufferedStream.

Reply:
Can you share its advantages over MemoryStream?

Reply:
Thanks, Shivangi!  I'll have a quick look, and consider this as a possible improvement for a later PR.  If you're able to show how I could make concrete performance improvements with this approach, that would help accelerate this work.  Thanks!

Reply:
I did a small perf test between these 2 approaches but it gave me almost the same result.

Later I found [this](https://learn.microsoft.com/en-us/archive/blogs/brada/filestream-and-bufferedstream) doc:
>No, there is zero benefit from wrapping a BufferedStream around a FileStream. We copied BufferedStream’s buffering logic into FileStream about 4 years ago to encourage better default performance ([Pit of Success](https://learn.microsoft.com/en-us/archive/blogs/brada/the-pit-of-success)). **In fact, I don’t think there are any Streams in the .NET Framework that require it, but it might be needed by custom Stream implementations if they do not do buffering by default.**

Seems like all the streams in .NET framework do buffer by default. So in this case we don't need to wrap stream into BufferedStream.

Reply:
Thanks so much for the investigation, @ShivangiReja!